ksqldb-api-client/src/test/java/io/confluent/ksql/api/client/integration/ClientIntegrationTest.java

```diff
@@ -657,22 +679,111 @@ public void shouldInsertInto() throws Exception {
+  public void shouldPullQueryTableAfterInsert() throws Exception {
     // Given
     final KsqlObject insertRow = new KsqlObject()
-        .put("K", new KsqlObject().put("F1", new KsqlArray().add("my_key")))
+        .put("K", "my_key")
         .put("LONG", 11L);
 
     // When
+    client.insertInto(TEST_TABLE, insertRow).get();
+
+    // Then: cannot pull query table
+    final String query = "SELECT * from " + TEST_TABLE + " WHERE K='my_key';";
     final Exception e = assertThrows(
         ExecutionException.class, // thrown from .get() when the future completes exceptionally
-        () -> client.insertInto(AGG_TABLE, insertRow).get()
+        () -> client.executeQuery(query).get()
     );
 
     // Then
     assertThat(e.getCause(), instanceOf(KsqlClientException.class));
     assertThat(e.getCause().getMessage(), containsString("Received 400 response from server"));
-    assertThat(e.getCause().getMessage(), containsString("Cannot insert into a table"));
+    assertThat(e.getCause().getMessage(), containsString("table isn't queryable"));
+  }
+
+  @Test
+  public void shouldPushQueryTableAfterInsert() throws Exception {
+    // Given
+    final Map<String, Object> properties = new HashMap<>();
+    properties.put("auto.offset.reset", "earliest");
+
+    final KsqlObject insertRow = new KsqlObject()
+        .put("K", "my_key")
+        .put("LONG", 11L);
+
+    // When
+    final String query = "SELECT * from " + TEST_TABLE + " WHERE K='my_key' EMIT CHANGES LIMIT 1;";
+    StreamedQueryResult queryResult = client.streamQuery(query, properties).get();
+
+    final Row row = assertThatEventually(() -> {
+      // Potentially try inserting multiple times, in case the query wasn't started by the first time
+      try {
+        client.insertInto(TEST_TABLE, insertRow).get();
+      } catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+      return queryResult.poll(Duration.ofMillis(10));
+    }, is(notNullValue()));
+
+    // Then: a newly inserted row arrives
+    assertThat(row.getString("K"), is("my_key"));
+    assertThat(row.getLong("LONG"), is(11L));
+  }
+
+  @Test
+  public void shouldPullQueryCTASTableAfterInsert() throws Exception {
+    // Given
+    final KsqlObject insertRow = new KsqlObject()
+        .put("K", new KsqlObject().put("F1", new KsqlArray().add("my_key")))
+        .put("LONG", 11L);
+    final String key = "STRUCT(F1 := ARRAY['my_key'])";
+
+    // When
+    client.insertInto(AGG_TABLE, insertRow).get();
+
+    // Then: should NOT receive new row
+    final String query = "SELECT * from " + AGG_TABLE + " WHERE K=" + key + ";";
+    final List<Row> rows = client.executeQuery(query).get();
+
+    // Verify inserted row is as expected
+    // Pull query read from state store, inserting into CTAS table won't
+    // insert to state store
+    assertThat(rows, hasSize(0));
+  }
+
+  @Test
+  public void shouldPullQueryCTASTableAfterInsertStream() throws Exception {
```

Comment:
`insert into stream` did always work, so it seems the test is actually not related to this PR?

Reply:
Want to test inserting to stream which backs CTAS table and it will be inserted to table. I'm trying to cover all possible ways to insert to a table.

Reply:
If this closes a test gap, great. I was just wondering if this case was already covered or not.

Reply:
Ditto to my comment above -- I'm not convinced this test belongs in this file. I definitely think we should have a test that verifies that INSERT VALUES is allowed for tables, but I don't think we need to test auxiliary behaviors.

Reply:
I can remove this as well